File: python/plusultra/loader.py
"""Model loading and initialization."""

import os
import json
import hashlib
from pathlib import Path

try:
    import onnxruntime as ort
except ImportError:
    raise ImportError(
        "onnxruntime is required. Install with: pip install onnxruntime"
    )

from .agent import Agent


def verify_model_integrity(model_dir: str) -> bool:
    """Verify model integrity using SHA-256 hashes from manifest.json."""
    manifest_path = Path(model_dir) / "manifest.json"

    if not manifest_path.exists():
        raise FileNotFoundError(f"manifest.json not found in {model_dir}")

    with open(manifest_path, "r") as f:
        manifest = json.load(f)

    for file_info in manifest.get("files", []):
        file_path = Path(model_dir) / file_info["path"]
        expected_hash = file_info["sha256"]

        if not file_path.exists():
            print(f"⚠️  Missing: {file_path}")
            continue

        # Compute SHA-256
        sha256_hash = hashlib.sha256()
        with open(file_path, "rb") as f:
            for chunk in iter(lambda: f.read(4096), b""):
                sha256_hash.update(chunk)

        actual_hash = sha256_hash.hexdigest()
        if actual_hash != expected_hash:
            raise ValueError(
                f"Hash mismatch for {file_info['path']}: "
                f"expected {expected_hash}, got {actual_hash}"
            )

        print(f"✓ {file_info['path']}")

    return True


def load(model_dir: str = "../model", verify_hash: bool = True) -> Agent:
    """
    Load a plusultra model from disk.

    Args:
        model_dir: Path to model directory (contains weights and tokenizer)
        verify_hash: If True, verify SHA-256 hashes against manifest.json

    Returns:
        Agent: Ready-to-use inference agent

    Example:
        agent = load("./model")
        result = agent.predict("text", {...})
    """
    model_dir = Path(model_dir).resolve()

    if not model_dir.exists():
        raise FileNotFoundError(f"Model directory not found: {model_dir}")

    print(f"📦 Loading model from {model_dir}")

    # Verify integrity
    if verify_hash:
        print("🔐 Verifying SHA-256 hashes...")
        verify_model_integrity(str(model_dir))

    # Initialize ONNX Runtime session
    model_file = model_dir / "model.onnx"
    if not model_file.exists():
        # Try to find .onnx file
        onnx_files = list(model_dir.glob("*.onnx"))
        if not onnx_files:
            raise FileNotFoundError(
                f"No .onnx model file found in {model_dir}. "
                f"Export from HuggingFace with: "
                f"python -m transformers.models.export onnx ..."
            )
        model_file = onnx_files[0]

    print(f"🚀 Initializing ONNX Runtime session...")
    session = ort.InferenceSession(
        str(model_file),
        providers=["CPUExecutionProvider"]
    )

    # Load tokenizer
    tokenizer_path = model_dir / "tokenizer.json"
    if not tokenizer_path.exists():
        raise FileNotFoundError(f"tokenizer.json not found in {model_dir}")

    print(f"✓ Model loaded successfully")

    return Agent(session, model_dir, tokenizer_path)

File: rust/src/loader.rs
//! Model loading and initialization

use anyhow::{anyhow, Result};
use sha2::{Sha256, Digest};
use std::fs;
use std::path::Path;
use crate::Agent;

/// Verify model integrity using SHA-256 hashes from manifest.json
pub fn verify_model_integrity(model_dir: &Path) -> Result<()> {
    let manifest_path = model_dir.join("manifest.json");

    if !manifest_path.exists() {
        return Err(anyhow!("manifest.json not found in {:?}", model_dir));
    }

    let manifest_content = fs::read_to_string(&manifest_path)?;
    let manifest: serde_json::Value = serde_json::from_str(&manifest_content)?;

    if let Some(files) = manifest.get("files").and_then(|v| v.as_array()) {
        for file_info in files {
            let path = file_info
                .get("path")
                .and_then(|v| v.as_str())
                .ok_or_else(|| anyhow!("Invalid manifest format"))?;

            let expected_hash = file_info
                .get("sha256")
                .and_then(|v| v.as_str())
                .ok_or_else(|| anyhow!("Missing hash for {}", path))?;

            let file_path = model_dir.join(path);

            if !file_path.exists() {
                println!("⚠️  Missing: {}", path);
                continue;
            }

            // Compute SHA-256
            let file_bytes = fs::read(&file_path)?;
            let mut hasher = Sha256::new();
            hasher.update(&file_bytes);
            let hash = hex::encode(hasher.finalize());

            if hash != expected_hash {
                return Err(anyhow!(
                    "Hash mismatch for {}: expected {}, got {}",
                    path, expected_hash, hash
                ));
            }

            println!("✓ {}", path);
        }
    }

    Ok(())
}

/// Load a plusultra model from disk
///
/// # Arguments
///
/// * `model_dir` - Path to model directory (contains model.onnx and tokenizer.json)
/// * `verify_hash` - If true, verify SHA-256 hashes against manifest.json
///
/// # Example
///
/// ```ignore
/// let agent = load("./model", true)?;
/// ```
pub fn load(model_dir: &str, verify_hash: bool) -> Result<Agent> {
    let model_path = Path::new(model_dir);

    if !model_path.exists() {
        return Err(anyhow!("Model directory not found: {:?}", model_dir));
    }

    println!("📦 Loading model from {:?}", model_path);

    // Verify integrity
    if verify_hash {
        println!("🔐 Verifying SHA-256 hashes...");
        verify_model_integrity(model_path)?;
    }

    // Find ONNX model file
    let model_file = model_path.join("model.onnx");
    let model_file = if model_file.exists() {
        model_file
    } else {
        // Try to find any .onnx file
        let onnx_files: Vec<_> = fs::read_dir(model_path)?
            .filter_map(|e| {
                e.ok().and_then(|entry| {
                    let path = entry.path();
                    if path.extension().and_then(|s| s.to_str()) == Some("onnx") {
                        Some(path)
                    } else {
                        None
                    }
                })
            })
            .collect();

        if onnx_files.is_empty() {
            return Err(anyhow!(
                "No .onnx model file found in {:?}. \
                Export from HuggingFace with: \
                python -m transformers.models.export onnx ...",
                model_path
            ));
        }
        onnx_files[0].clone()
    };

    println!("🚀 Initializing ONNX Runtime session...");

    // Load tokenizer
    let tokenizer_path = model_path.join("tokenizer.json");
    if !tokenizer_path.exists() {
        return Err(anyhow!("tokenizer.json not found in {:?}", model_path));
    }

    println!("✓ Model loaded successfully");

    Agent::new(model_file, tokenizer_path)
}
